core/domain/stats_services.py

```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+            this will be equal to 'question_id' only.
+        entity_type: str. The type of entity i.e 'exploration' or 'question '.
+        interaction_id: str. The ID of the interaction.
+        change: LearnerAnswerDetailsChange. The change to be applied
+            to the LearnerAnswerDetails object.
+
+    Returns:
+        LearnerAnswerDetails. The resulting learner answer details object,
+            after applying the change.
+     """
+    learner_answer_details = get_learner_answer_details(
+        state_reference, entity_type)
+    if change.cmd == stats_domain.CMD_RECORD_LEARNER_ANSWER_INFO:
+        if not learner_answer_details:
+            learner_answer_details = stats_domain.LearnerAnswerDetails(
+                state_reference, entity_type, interaction_id, [], 0)
+        learner_answer_info_id = (
+            stats_domain.LearnerAnswerInfo.get_learner_answer_info_id(
+                state_reference, entity_type))
+        learner_answer_info = stats_domain.LearnerAnswerInfo(
+            learner_answer_info_id, change.answer,
+            change.answer_details, datetime.datetime.utcnow())
+        learner_answer_details.add_learner_answer_info(learner_answer_info)
+    elif change.cmd == stats_domain.CMD_DELETE_LEARNER_ANSWER_INFO:
+        learner_answer_details.delete_learner_answer_info(
+            change.learner_answer_info_id)
+    elif change.cmd == stats_domain.CMD_UPDATE_STATE_REFERENCE:
+        learner_answer_details.update_state_reference(change.new_value)
+    else:
+        raise Exception('Invalid change dict for learner answer details.')
+    return learner_answer_details
+
+
+def get_learner_answer_details_from_model(learner_answer_details_model):
+    """Returns a LearnerAnswerDetails domain object given a
+    LearnerAnswerDetailsModel loaded from the datastore.
+
+    Args:
+        learner_answer_details_model: LearnerAnswerDetailsModel. The learner
+            answer details model loaded from the datastore.
+
+    Returns:
+        LearnerAnswerDetails. A LearnerAnswerDetails domain object
+            corresponding to the given model.
+    """
+    return stats_domain.LearnerAnswerDetails(
+        learner_answer_details_model.state_reference,
+        learner_answer_details_model.entity_type,
+        learner_answer_details_model.interaction_id,
+        [stats_domain.LearnerAnswerInfo.from_dict(learner_answer_info_dict)
+         for learner_answer_info_dict
+         in learner_answer_details_model.learner_answer_info_list],
+        learner_answer_details_model.schema_version,
+        learner_answer_details_model.accumulated_answer_info_json_size_bytes)
+
+
+def get_learner_answer_details(state_reference, entity_type):
+    """Returns a LearnerAnswerDetails domain object, with given
+    state_reference, entity_type & interaction_id. This function checks
+    in the datastore if the corresponding LearnerAnswerDetailsModel exists,
+    if not then new domain object is created.
```

Comment:
This is not done, right? If not present, None is returned.

Reply:
Done